WIP: transaction control batch removal in local storage compaction

WillemKauf opened the pull request:
Opening now for CI.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

### Features

* Cool stuff

WillemKauf committed:
`storage`: add `self_compact_timestamp` to `index_state`

To mark the timestamp at which a `segment` is first self compacted.

This is going to be needed for the removal of transactional control
batches using `delete.retention.ms`.

WillemKauf committed:
`storage`: add `mark_segment_as_finished_self_compaction()`

Sets the `self_compact_timestamp` in the `index_state`, and
potentially flushes the `segment_index`.

WillemKauf committed:
`storage`: deprecate `segment::bitflags::finished_self_compaction`

And its related functions.

These bitflags are not persisted to disk and are only set in memory.

We now have the persisted information for the `self_compact_timestamp`
in `segment_index`, so we do not need the bitflag anymore.

WillemKauf committed:
`storage`: add `has_transaction_batches` to `index_state`

Answers the question of "does this segment have any transactional batches
in it?"

WillemKauf committed:
`storage`: add `tx_retention_ms` to `compaction_config`

WillemKauf committed:
`storage`: add `has_removable_transaction_batches()`

Returns `true` iff the `segment` in question has transaction batches,
and is past the transaction batch delete horizon.

Also rework the logic in the `tombstone_records()` function to closely
resemble the added logic in the new function (no functional change there).

WillemKauf committed:
`config`: rework `delete.retention.ms` validation

Because we will be using the configured value for `delete.retention.ms`
for the removal of transaction markers as well, we should no longer limit
the enablement of this topic/cluster property based on tiered storage
being disabled.

Safeguards are still in place in `ntp_config::tombstone_retention_ms()`
to ensure this value is not set if tiered storage properties are enabled.

Also update the cluster default value for `tombstone_retention_ms` to `24h`
(one day) to match Kafka's default.

WillemKauf committed:
`storage`: removing `ntp` stuff from compactible batch functions

WillemKauf committed:
`storage`: add `enhanced_key` utility functions

In order to get the `is_control` and `record_batch_type` fields back from
an `compacted_index::entry` which has been encoded via `enhance_key()` and
written to the `compacted_index` for a `segment`.

This will be used in a future commit to determine whether a record needs
to have its key appended to the `key_offset_map` built during sliding
window compaction.

WillemKauf committed:
`storage`: improve log line in `tx_reducer` by adding writer context

WillemKauf committed:
`storage`: remove transaction batches during compaction

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/66539-->
<!--aba487ab-4e63-4e7e-bde4-35578caca131-->
<!--919c4739-11b3-4a42-af03-ba97ff4bec60-->
<!--8e40fd3d-06a9-41df-b9b8-9cd71f3d2ac9-->
<!--771705c6-86ad-43ee-8dfd-c7cde223bd25-->
<!--76c1e79d-6a05-4465-a9b3-c70cb3524bd4-->
<!--81cbf0e3-7cd7-4cf6-919f-6ee37f7a7794-->
<!--7ff99628-1f7d-483b-b1a1-1c05cbbaf4e3-->
<!--title-->
## Retry command for [Build#66539](https://buildkite.com/redpanda/redpanda/builds/66539)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/delete_retention_ms_test.py::DeleteRetentionMsTest.test_get_cluster_config
tests/rptest/tests/controller_snapshot_test.py::ControllerSnapshotTest.test_upgrade_compat
tests/rptest/tests/delete_retention_ms_test.py::DeleteRetentionMsTest.test_alter_topic_config
tests/rptest/tests/topic_creation_test.py::TopicAutocreateTest.topic_autocreate_test
tests/rptest/tests/describe_topics_test.py::DescribeTopicsTest.test_describe_topics_with_documentation_and_types
tests/rptest/transactions/compaction_e2e_test.py::CompactionWithRecoveryTest.test_tx_compaction_with_recovery
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":false,"mixed_versions":true,"with_chunked_compaction":true,"with_iceberg":false}
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy@{"cloud_storage_type":2,"short_retention":true}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--e0a2b872-c28f-438e-9f8e-c12e0d4f05f0-->
<!--TEST RESULTS-->
<!--aa23bdaa-8e30-416e-adc8-7af64202632d-->
<!--TEST RESULTS-->
<!--f068d971-cc04-4eb6-b5ca-fa3ec96e18b1-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66539">build#66539</a></summary>


| test_class                  | test_method                                       | test_arguments                                                                                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:--------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ControllerSnapshotTest      | test_upgrade_compat                               |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66539#01971a35-ac8c-4307-999d-7c7b51b26a15 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DeleteRetentionMsTest       | test_alter_topic_config                           |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66539#01971a35-ac8d-4405-ab00-845fe463596f | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DeleteRetentionMsTest       | test_alter_topic_config                           |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66539#01971a4f-81d7-4855-9d00-035fe10fe5c7 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DeleteRetentionMsTest       | test_get_cluster_config                           |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66539#01971a35-ac8e-46f8-8fba-73890b91feaa | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DeleteRetentionMsTest       | test_get_cluster_config                           |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66539#01971a4f-81d5-46b9-bb16-be116fc84983 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DescribeTopicsTest          | test_describe_topics_with_documentation_and_types |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66539#01971a35-ac8d-410a-ba3d-a81ed995bcb8 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| ShadowIndexingWhileBusyTest | test_create_or_delete_topics_while_busy           | {"cloud_storage_type": 2, "short_retention": true}                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66539#01971a4f-81d6-4616-9c66-0c92d4b5ca7c | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                 |
| RandomNodeOperationsTest    | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": true, "with_chunked_compaction": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66539#01971a4f-81d5-46b9-bb16-be116fc84983 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                 |
| RandomNodeOperationsTest    | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66539#01971a4f-81d5-46b9-bb16-be116fc84983 | FLAKY         | 20/21    | upstream reliability is '97.72727272727273'. current run reliability is '95.23809523809523'. drift is 2.48918 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest    | test_node_operations                              | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": true, "with_chunked_compaction": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66539#01971a4f-81d6-4616-9c66-0c92d4b5ca7c | FLAKY         | 14/18    | upstream reliability is '99.2248062015504'. current run reliability is '77.77777777777779'. drift is 21.44703 and the allowed drift is set to 50. The test should PASS |
| TopicAutocreateTest         | topic_autocreate_test                             |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66539#01971a35-ac8d-4405-ab00-845fe463596f | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| CompactionWithRecoveryTest  | test_tx_compaction_with_recovery                  |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66539#01971a4f-81d7-4855-9d00-035fe10fe5c7 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66803">build#66803</a></summary>


| test_class                 | test_method                                       | test_arguments                                                                                                                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:---------------------------|:--------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| ControllerSnapshotTest     | test_upgrade_compat                               |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739d0-85f0-4078-a4d7-99b747cf617a | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| ControllerSnapshotTest     | test_upgrade_compat                               |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739f5-0b8a-46b3-95b4-ce108c6ee14d | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| DatalakeBatchingTest       | test_batching                                     | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "expect_large_files": true, "query_engine": "spark"}                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739f5-0b88-4e11-9f86-439527892d89 | FLAKY         | 20/21    |                                                                                                                                                           |
| DeleteRetentionMsTest      | test_alter_topic_config                           |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739d0-85f0-4078-a4d7-99b747cf617a | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| DeleteRetentionMsTest      | test_alter_topic_config                           |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739f5-0b89-4d15-8fac-eeae2a181ac0 | FAIL          | 0/21     |                                                                                                                                                           |
| DeleteRetentionMsTest      | test_get_cluster_config                           |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739d0-85f1-4727-b1a0-a2ce0e9c5791 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| DeleteRetentionMsTest      | test_get_cluster_config                           |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739f5-0b8a-46b3-95b4-ce108c6ee14d | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| DescribeTopicsTest         | test_describe_topics_with_documentation_and_types |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739d0-85f2-4588-9d1f-c38b11399326 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| DescribeTopicsTest         | test_describe_topics_with_documentation_and_types |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739f5-0b88-4e11-9f86-439527892d89 | FAIL          | 0/21     |                                                                                                                                                           |
| EndToEndShadowIndexingTest | test_write                                        | {"cloud_storage_type": 1}                                                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739d0-85f0-4078-a4d7-99b747cf617a | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest   | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": true, "with_chunked_compaction": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739f5-0b88-4e11-9f86-439527892d89 | FAIL          | 0/2      |                                                                                                                                                           |
| RandomNodeOperationsTest   | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": true, "with_chunked_compaction": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739f5-0b8a-46b3-95b4-ce108c6ee14d | FAIL          | 0/2      | The test has failed across all retries                                                                                                                    |
| RandomNodeOperationsTest   | test_node_operations                              | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": true, "with_chunked_compaction": true, "with_iceberg": false}   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739f5-0b89-4d15-8fac-eeae2a181ac0 | FAIL          | 0/4      |                                                                                                                                                           |
| RandomNodeOperationsTest   | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": true, "with_chunked_compaction": true, "with_iceberg": false}   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739f5-0b8a-46b3-95b4-ce108c6ee14d | FAIL          | 0/3      | The test has failed across all retries                                                                                                                    |
| TopicAutocreateTest        | topic_autocreate_test                             |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739d0-85f0-4078-a4d7-99b747cf617a | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| TopicAutocreateTest        | topic_autocreate_test                             |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739f5-0b88-4e11-9f86-439527892d89 | FAIL          | 0/21     |                                                                                                                                                           |
| UpgradeBackToBackTest      | test_upgrade_with_all_workloads                   | {"single_upgrade": false}                                                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739f5-0b89-4d15-8fac-eeae2a181ac0 | FLAKY         | 4/5      |                                                                                                                                                           |
| CompactionWithRecoveryTest | test_tx_compaction_with_recovery                  |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66803#019739f5-0b89-4d15-8fac-eeae2a181ac0 | FAIL          | 0/5      |                                                                                                                                                           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66963">build#66963</a></summary>


| test_class                       | test_method                                       | test_arguments                                                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------------|:--------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudRetentionTest               | test_cloud_retention                              | {"cloud_storage_type": 2, "max_consume_rate_mb": null}                                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| ClusterConfigAzureSharedKey      | test_live_shared_key_change                       | {"cloud_storage_type": 2}                                                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| DescribeTopicsTest               | test_describe_topics_with_documentation_and_types |                                                                                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#0197433f-919d-47fd-96b3-326a6dd30bb8 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DescribeTopicsTest               | test_describe_topics_with_documentation_and_types |                                                                                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e6e-4570-87d4-46ff500b796e | FAIL          | 0/21     |                                                                                                                                                                        |
| EndToEndTopicRecovery            | test_restore                                      | {"cloud_storage_type": 2, "message_size": 5000, "num_messages": 100000, "recovery_overrides": {}}                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| EndToEndTopicRecovery            | test_restore_with_aborted_tx                      | {"cloud_storage_type": 2, "recovery_overrides": {}}                                                                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| EndToEndTopicRecovery            | test_restore_with_config_batches                  | {"cloud_storage_type": 2, "num_messages": 2}                                                                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest         | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": true, "with_chunked_compaction": true, "with_iceberg": false}                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e6e-4570-87d4-46ff500b796e | FLAKY         | 1/4      |                                                                                                                                                                        |
| RandomNodeOperationsTest         | test_node_operations                              | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": false}                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest         | test_node_operations                              | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false}                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e72-49ef-905b-420f63436fbd | FLAKY         | 10/13    |                                                                                                                                                                        |
| RandomNodeOperationsTest         | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false}                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e6e-4570-87d4-46ff500b796e | FLAKY         | 4/5      |                                                                                                                                                                        |
| RandomNodeOperationsTest         | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": true, "with_chunked_compaction": false, "with_iceberg": false}                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e71-4b1f-833d-8dce6fba08f3 | FLAKY         | 10/21    |                                                                                                                                                                        |
| RandomNodeOperationsTest         | test_node_operations                              | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": true, "with_chunked_compaction": true, "with_iceberg": false}                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e72-49ef-905b-420f63436fbd | FLAKY         | 4/12     |                                                                                                                                                                        |
| RandomNodeOperationsTest         | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": true, "with_chunked_compaction": true, "with_iceberg": false}                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e6e-4570-87d4-46ff500b796e | FLAKY         | 3/4      |                                                                                                                                                                        |
| DisablingPartitionsTest          | test_disable                                      |                                                                                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#0197433f-919a-46d7-be0e-5c1348bdc295 | FLAKY         | 18/21    | upstream reliability is '86.73913043478261'. current run reliability is '85.71428571428571'. drift is 1.02484 and the allowed drift is set to 50. The test should PASS |
| ShadowIndexingCompactedTopicTest | test_upload                                       | {"cloud_storage_type": 2}                                                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| TieredStorageTest                | test_tiered_storage                               | {"cloud_storage_type_and_url_style": [2, "virtual_host"], "test_case": {"name": "(TS_Read == True, AdjacentSegmentMergerReupload == True, SpilloverManifestUploaded == True)"}} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| TieredStorageTest                | test_tiered_storage                               | {"cloud_storage_type_and_url_style": [2, "virtual_host"], "test_case": {"name": "(TS_Read == True, TS_Timequery == True)"}}                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| TopicRecoveryTest                | test_admin_api_recovery                           | {"cloud_storage_type": 2}                                                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| TopicRecoveryTest                | test_fast1                                        | {"cloud_storage_type": 2}                                                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| TopicRecoveryTest                | test_fast3                                        | {"cloud_storage_type": 2}                                                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| TopicRecoveryTest                | test_missing_partition                            | {"cloud_storage_type": 2}                                                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| TopicRecoveryTest                | test_missing_topic_manifest                       | {"cloud_storage_type": 2}                                                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| TopicRecoveryTest                | test_prevent_recovery                             | {"cloud_storage_type": 2}                                                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| TopicRecoveryTest                | test_time_based_retention                         | {"cloud_storage_type": 2}                                                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e70-4873-8273-7f157d705a87 | FAIL          | 0/1      |                                                                                                                                                                        |
| CompactionWithRecoveryTest       | test_tx_compaction_with_recovery                  |                                                                                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66963#01974357-6e72-49ef-905b-420f63436fbd | FAIL          | 0/21     |                                                                                                                                                                        |
</details>
<!--/body-->

WillemKauf committed:
`storage`: add `max_clean_and_removable_offset()`

Demarcates { SECTION A | SECTION B } in a compacted log as discussed.

WillemKauf committed:
`raft`: add `recovery_start_time` to `follower_index_metadata`

To track the start time of a recovery for a learner from this leader.

WillemKauf committed:
`cluster`: add `raft_recovery_disable_compaction_safety_checks`

Escape hatch to force unsafe recovery, when the user:
a) has encountered wide-spread node instability and frequent recovery
   restarts due to these conditions
b) does not care about possibly diverged replica state w/r/t tombstones
   & transactions
c) is ENTIRELY CONFIDENT that a diverged replica state is not going to
   happen.

The same thing could be achieved by disabling `delete.retention.ms`, but
it feels nicer to have an explicit toggle for this.

WillemKauf committed:
`raft`: add `recovery_rpc_handler/types`

Three RPC commands for leader/learner message passing as previously
discussed:

1. `get_learner_state()`
2. `reset_learner()`
3. `commit_new_leader_id()`

For use in the `recovery_stm` to avoid any race conditions with
`delete.retention.ms` enabled `compact` topics.

WillemKauf committed:
`redpanda`: add `recovery_rpc_handler` to `runtime_services`

WillemKauf committed:
`raft`: add `ss::sharded<rpc::connections_cache>&` holder to `consensus`/`group_manager`

The `recovery_rpc_handler` will use a raw `rpc::connections_cache` to issue
RPCs between leaders and learners during recovery, instead of the buffered
protocols typically seen in `raft`.

These RPCs should be _very_ infrequent in comparison, so there is no
real need for a buffered protocol when issuing these calls.

WillemKauf committed:
`raft`: add `ss::sharded<rpc::connection_cache>&` holder to `recovery_stm`

The `recovery_stm` will be the user issuing `recovery_rpc` commands via the
connection cache.

WillemKauf committed:
`raft`: add `consensus::clear_state()`

This will be the function invoked to fully reset a learner's state
when it is possible that it will diverge during a recovery.

WillemKauf committed:
`raft`: safe recovery in the presence of `delete.retention.ms`

Adding all the new functionality that is required for safe recovery
in the presence of a `compact` topic with `delete.retention.ms` enabled.

WillemKauf commented:
/ci-repeat 1

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/66778-->
<!--d321fabf-8070-4d86-892b-1a560d7b42bb-->
<!--592cf22b-e51a-4a88-85bb-99da84a65dfe-->
<!--title-->
## Retry command for [Build#66778](https://buildkite.com/redpanda/redpanda/builds/66778)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/partition_move_interruption_test.py::PartitionMoveInterruption.test_cancelling_partition_move_x_core@{"compacted":false,"recovery":"no_recovery","replication_factor":3,"unclean_abort":false}
tests/rptest/tests/partition_move_interruption_test.py::PartitionMoveInterruption.test_cancelling_partition_move_x_core@{"compacted":false,"recovery":"no_recovery","replication_factor":3,"unclean_abort":true}
tests/rptest/tests/scaling_up_test.py::ScalingUpTest.test_adding_nodes_to_cluster@{"partition_count":1}
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_partition_autobalancer_sanction@{"disable_license":true}
tests/rptest/tests/scaling_up_test.py::ScalingUpTest.test_topic_hot_spots
tests/rptest/tests/datalake/recovery_mode_test.py::DatalakeRecoveryModeTest.test_recovery_mode@{"catalog_type":"rest_hadoop","cloud_storage_type":1}
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_decommissioning_finishes_after_manual_cancellation@{"delete_topic":true}
tests/rptest/tests/full_node_recovery_test.py::FullNodeRecoveryTest.test_node_recovery@{"recovery_type":"partial"}
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_flipping_decommission_recommission@{"node_is_alive":false}
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_decommissioning_working_node@{"delete_topic":false,"tick_interval":3600000}
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_recommissioning_do_not_stop_all_moves_node
tests/rptest/tests/rack_aware_replica_placement_test.py::RackAwarePlacementTest.test_rack_awareness_after_node_operations@{"rack_layout":"ABCDEF"}
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_recycle_all_nodes@{"auto_assign_node_id":false}
tests/rptest/tests/partition_reassignments_test.py::PartitionReassignmentsTest.test_reassignments
tests/rptest/tests/shard_placement_test.py::ShardPlacementTest.test_node_join@{"disable_license":false}
tests/rptest/transactions/tx_feature_flag_test.py::TxFeatureFlagTest.test_disabling_transactions_after_they_being_used
tests/rptest/tests/delete_records_test.py::DeleteRecordsTest.test_delete_records_concurrent_truncations@{"cloud_storage_enabled":true,"truncate_point":"at_high_watermark"}
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTestWithDisruptions.test_write_with_node_failures@{"cloud_storage_type":1}
tests/rptest/tests/partition_movement_test.py::PartitionMovementTest.test_dynamic@{"num_to_upgrade":2}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":null,"use_alias":true},"transfer_leadership":true}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"prepared"},"use_alias":true},"transfer_leadership":false}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"executing"},"use_alias":true},"transfer_leadership":true}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":true},"transfer_leadership":true}
tests/rptest/tests/metrics_reporter_test.py::MultiNodeMetricsReporterTest.test_redpanda_metrics_reporting
tests/rptest/tests/controller_erase_test.py::ControllerEraseTest.test_erase_controller_log@{"partial":false}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_creating_and_listing_migrations_wo_license
tests/rptest/tests/data_transforms_test.py::DataTransformsTest.test_patch@{"use_rpk":false}
tests/rptest/tests/prefix_truncate_recovery_test.py::PrefixTruncateRecoveryTest.test_prefix_truncate_recovery@{"acks":-1,"start_empty":true}
tests/rptest/tests/delete_retention_ms_test.py::DeleteRetentionMsTest.test_get_cluster_config
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":false,"mixed_versions":true,"with_chunked_compaction":true,"with_iceberg":false}
tests/rptest/tests/controller_snapshot_test.py::ControllerSnapshotTest.test_upgrade_compat
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/66803-->
<!--579e77b4-c316-44cd-82a5-766159515164-->
<!--e452e90c-c604-43bf-8c77-6f715be2285e-->
<!--057a699e-f120-4fee-91b4-6ee2771f3ec2-->
<!--601e6642-3c95-412f-b7fa-3e5be7775d66-->
<!--title-->
## Retry command for [Build#66803](https://buildkite.com/redpanda/redpanda/builds/66803)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/delete_retention_ms_test.py::DeleteRetentionMsTest.test_get_cluster_config
tests/rptest/tests/describe_topics_test.py::DescribeTopicsTest.test_describe_topics_with_documentation_and_types
tests/rptest/tests/delete_retention_ms_test.py::DeleteRetentionMsTest.test_alter_topic_config
tests/rptest/tests/topic_creation_test.py::TopicAutocreateTest.topic_autocreate_test
tests/rptest/tests/controller_snapshot_test.py::ControllerSnapshotTest.test_upgrade_compat
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":false,"mixed_versions":true,"with_chunked_compaction":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":true,"mixed_versions":true,"with_chunked_compaction":true,"with_iceberg":false}
```
<!--/body-->

WillemKauf committed:
`raft`: add `consensus::truncate_state()` function

NFC, adds a new function `truncate_state()` to replace some code
which will be re-used in a later commit.

WillemKauf committed:
`rptest`: add `RaftRecoveryWithCompactionTest`

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/66963-->
<!--f5c3b84e-dd54-4b79-8860-b90c89a7c732-->
<!--title-->
## Retry command for [Build#66963](https://buildkite.com/redpanda/redpanda/builds/66963)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/describe_topics_test.py::DescribeTopicsTest.test_describe_topics_with_documentation_and_types
```
<!--/body-->

WillemKauf committed:
`raft`: add `recovery_rpc_protocol`

Class which uses PIMPL to perform client related operations through
the `recovery_rpc_handler/service`.

WillemKauf committed:
`raft`: add `recovery_rpc_client.cc/.h`

Implementation of `recovery_rpc_protocol::impl`. Uses the `rpc::connection_cache`
and `recovery_rpc_client_protocol` to issue RPCs to the `recovery_rpc_handler`.

WillemKauf committed:
`raft`: add `consensus_accessor`

WillemKauf committed:
raft recovery test

WillemKauf committed:
`raft`: add `get_log_truncations()` to `probe`

WillemKauf committed:
`storage`: add `control_batches_discarded` to `copy_data_segment_reducer::stats`

To help track removed control batches during compaction.

WillemKauf committed:
tx compaction utils enhancements

micheleRP commented on src/v/config/configuration.cc:
```diff
@@ -1257,6 +1255,16 @@ configuration::configuration()
+  , raft_recovery_disable_compaction_safety_checks(
+      *this,
+      "raft_recovery_disable_compaction_safety_checks",
+      "Disables safety checks in the Raft recovery subsystem which are used to "
+      "ensure divergence-free states when recovering from a compacted topic. "
+      "This will prevent redpanda from forcing followers from resetting their "
+      "log during a possibly unsafe recovery, but can lead to diverged "
+      "replica state. Normally, this option should be disabled.",
```
```suggestion
      "Disables the safety checks in the Raft recovery subsystem used to "
      "ensure divergence-free states when recovering from a compacted topic. "
      "This prevents Redpanda from forcing followers from resetting their "
      "log during a possibly unsafe recovery. Because it can lead to diverged "
      "replica state, this option generally should be disabled.",
```
just a style suggestion, if it makes sense

WillemKauf committed:
Temp
